experimental/offline_attestation/README.md

```diff
@@ -0,0 +1,111 @@
+  by the evidence that only appropriate enclaves would have access to the
+  private key) to encrypt requests to the server. Clients also provide
+  additional information (e.g. the client's public key) in the request to allow
+  the server to encrypt the response.
+
+- This approach has some advantages compared to the interactive handshake-based
+  approach:
+  - Reduced bandwith and latency
+  - No need for sticky sessions in a load-balanced environment
+- And some disadvantages:
+  - No forward secrecy since the same public key is used for encrypting all
+    requests
+  - No built-in replay protection
+
+## Implementation Details
+
+This experimental example provides simplified client and server implementations.
+
+### Server
+
+The server generates an asymmetric key-pair at startup. It then generates an
+attestation report that binds the public key to the enclave hardware and the
+version of the server code. For this example we just generate a fake empty
+attestation report that is always accepted by the client.
+
+For more realistic implementations that support load-balancing a
+key-provisioning mechanism is required to ensure that all instances of the
+server have access to the private key. This could be done in two ways:
+
+- A separate key management server (KMS): The KMS would generate the key-pair
+  and attestation report. When a server instance starts up it sets up a
+  mutually-attested connection with the KMS and requests the private key. The
+  KMS is then responsible for validating that the server is running an
+  appropriate version of the software in an appropriate enclave before sharing
+  the private key. The attestation report will include evidence about the KMS,
+  not the server, so the KMS code would need to be reviewed to ensure that it
+  will do the server validation correctly.
+- Key-sharing between servers: The first server to start up will be instructed
+  to generate a key-pair and attestation reports. When other servers start up,
+  they will be instructed to connect to an existing server to get the private
+  key via a mutually-attested connection. This approach might present some
+  challenges for key rotation or upgrading to new versions of the server code.
+
+Once the server has access to the private key it can start serving requests. It
+decrypts the ciphertext from each request. The request also includes a public
+key for the client. The server processes the cleartext, encrypts it with the
+client's public key, and returns the encrypted response.
+
+### Client
+
+For illustration purposes the client fetches the public key and associated
+attestation report from the server and validates the report. In a more realistic
+implemenation the retrieval and validation of the public key and attestation
+report would happen out-of-band. The public key would then be provided to the
+client app once the validity of the key and attestation report has been
```

Comment:
Will this key be hardcoded into the app? If so, how would we work with revocations?

Reply:
It could be hardcoded or done via configuration. Revocations would be tricky. One option would be to deploy new versions of the client with new keys. As this roll-out will likely take time it will be a challenge to ensure that all the old clients still connect to server instances that have the old key and all new clients connect to servers with the new key. This could be done by creating a new versioned URL every time the key is rotated. Another option is to allow servers to have multiple keys and then stop using the old ones when all the clients have been upgraded. Neither of these are great, but I can't think of many better options.